### PR TITLE
Allow different name among the bundles

### PR DESCRIPTION
**Pull request by tomahawk28**
Issue found during https://github.com/giantswarm/app-operator/pull/11

### CODE REVIEW

File: bundles.go [hunk 1]
```diff
@@ -54,7 +47,7 @@ func (b Bundles) hasDuplicatedVersions() bool {
 		var seen int
 
 		for _, b2 := range b {
-			if b1.Version == b2.Version && b1.Provider == b2.Provider {
+			if b1.Version == b2.Version && b1.Provider == b2.Provider && b1.Name == b2.Name {
```
Comment (rossf7) [inline bundles.go]: Change makes sense but I don't know how to test this. The issue is whether it causes problems with the changelogs we display to customers. @tuommaki WDYT?

Comment (tomahawk28) [reply bundles.go]: Hmm, I don't think changelogs would affect this problem.

**Comment by tomahawk28** [comment]
I'm blocking by this, will continue after merging it.

File: bundles.go (hunk 1 above)
Comment (rossf7) [reply bundles.go]: This library is used to format the changelogs we show in gsctl and Happa.

e.g.

```
gsctl list releases
gsctl show release 6.2.0
```

**Comment by tomahawk28** [comment]
I understood that we may have a problem during release this change, I got your point. @tuommaki. 
Can we work with this change until VOO is ready? Since I would like to move app catalog epic forward, this change is desperately needed.

**Comment by rossf7** [comment]
@tuommaki Adding a new test case for what we need to do in app-operator makes sense to me. Where are the tests for the release computation? Sorry I don't know this part of our codebase well.

Also as @tomahawk28 says this is blocking us on app-operator dev. This change could break in the future but its also going away with VOO. cc @puja108

**Comment by tuommaki** [comment]
Relevant tests are here:
* https://github.com/giantswarm/versionbundle/blob/master/release_test.go
* https://github.com/giantswarm/versionbundle/blob/master/index_release_test.go

If we introduce issues here there are two problem groups:
1. Cosmetic issues where e.g. release changelog is invalid and "only" causes some confusion among customers. Solution engineers would need to then explain these issues to customers one-by-one. Something that possibly could be done, but what would be quite painful if it lasts long.
2. :warning: Release computation breaks and cluster component composition is not what we expect. This could be potentially disastrous if wrong combination of operators get combined for some specific cluster. In best case this only prevents cluster to run and kind of "locks" the cluster in upgrades and prevents creation of new clusters. In worst case this makes operators to "play against each other" which can then cause disruption to customer workloads by e.g. taking ingresses down or running out of resources, using wrong networking components or similar. This would be really bad because I can imagine it to take hours to fix this sort of situation when it really goes bad.

These are the reasons I think this should be tested carefully. The workaround being a hack doesn't really matter to me by itself when this code will soon go away anyway, but we must be careful to not break clusters.

**Comment by kopiczko** [comment]
JFYI: This repo is gonna be retired with VOO. I think less time you sped on that is better even if the solution is ugly.

Disclaimer: I don't follow this PR.

**Comment by rossf7** [comment]
I'd say we should add cases to the tests mentioned by @tuommaki  and then go with this change. We need to proceed with app-operator for VOO so we can retire this library.

**Comment by tomahawk28** [comment]
Absolutely, I'll try to add testing there.

**Comment by tomahawk28** [comment]
@tuommaki So should I  add more tests about `bundles with the different name and same version` in `CompileRelease` and `NewRelease` function?

**Comment by tuommaki** [comment]
> So should I add more tests about bundles with the different name and same version in CompileRelease and NewRelease function?

I think easiest approach would be to change existing version bundle sets (the more complicated ones per test) by adding app-operator bundles there as they would be expected to be exposed (i.e. with name & version you are about to create). I don't think there's need to create new test cases as such.

I think these tests should cover the important bits:
* https://github.com/giantswarm/versionbundle/blob/master/release_test.go#L573
* https://github.com/giantswarm/versionbundle/blob/master/index_release_test.go#L25
* https://github.com/giantswarm/versionbundle/blob/master/index_release_test.go#L1054
* https://github.com/giantswarm/versionbundle/blob/master/index_release_test.go#L1603
* https://github.com/giantswarm/versionbundle/blob/master/index_release_test.go#L1711

**Comment by tomahawk28** [comment]
> by adding app-operator bundles there as they would be expected to be exposed (i.e. with name & version you are about to create

Thanks, I got it

File: index_release_test.go [hunk 2]
```diff
@@ -36,6 +36,16 @@ func Test_buildReleases(t *testing.T) {
 				{
 					Active: true,
 					Authorities: []Authority{
+						{
+							Endpoint: urlMustParse("http://app-operator:8000/"),
+							Name:     "app-operator",
+							Version:  "0.1.0",
+						},
+						{
+							Endpoint: urlMustParse("http://appcatalog-operator:8000/"),
```
Comment (rossf7) [inline index_release_test.go]: I don't get the `appcatalog-operator`? We're not going to create this operator. app-operator needs to have 2 controllers.

If you need another operator for the test then picking an existing operator is less confusing IMO.

**Comment by tomahawk28** [comment]
@tuommaki I only added `Bundle` and `Changelog` wherein a unittest means to test multiple `Bundle`. In my PR in app-operator, I only add two different `Bundle` so I don't really see a problem with `Release` which I didn't mean to disrupt.

File: index_release_test.go (hunk 2 above)
Comment (tomahawk28) [reply index_release_test.go]: oops, I will change the name as `app-controller` and `appcatalog-controller` then.

**Comment by tomahawk28** [comment]
```#!/bin/bash -eo pipefail
wget -q $(curl -sS -H "Authorization: token $RELEASE_TOKEN" https://api.github.com/repos/giantswarm/architect/releases/latest | grep browser_download_url | head -n 1 | cut -d '"' -f 4)
chmod +x ./architect
./architect version
```
Some error from circle CI

**Comment by rossf7** [comment]
@tomahawk28 Yes CI is broken due to GitHub rate limiting. You can check on progress in #sig-releng channel.

File: index_release_test.go [hunk 3]
```diff
@@ -36,6 +36,16 @@ func Test_buildReleases(t *testing.T) {
 				{
 					Active: true,
 					Authorities: []Authority{
+						{
+							Endpoint: urlMustParse("http://app-controller:8000/"),
+							Name:     "app-controller",
+							Version:  "0.1.0",
+						},
+						{
+							Endpoint: urlMustParse("http://appcatalog-controller:8000/"),
```
Comment (tuommaki) [inline index_release_test.go]:
```suggestion
							Endpoint: urlMustParse("http://app-controller:8000/"),
```

The URL must be the same when the running process/POD is the same.

**Comment by tomahawk28** [comment]
Thanks, To be sure I would definitely test against our cluster.

**Pull request merged**